// atft/esm5/lib/object/text/text-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractLazyObject3D } from '../abstract-lazy-object-3d';
import { appliedMaterial } from '../../util';
import { fixCenter } from '../../util/fix-center';
import { RendererService } from '../../renderer/renderer.service';
var TextMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TextMeshComponent, _super);
    function TextMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.material = 'basic';
        _this._materialColor = 0xDADADA;
        _this._text = 'Text';
        _this.size = 10;
        _this.height = 0.3;
        _this.curveSegments = 2;
        _this.bevelEnabled = false;
        _this.bevelThickness = 0.1;
        _this.bevelSize = 0.1;
        _this.bevelOffset = 0;
        _this.bevelSegments = 1;
        _this.fontUrl = './assets/font/helvetiker_regular.typeface.json';
        _this.castShadow = true;
        _this.receiveShadow = true;
        _this.depthWrite = true;
        _this.centered = true;
        return _this;
    }
    Object.defineProperty(TextMeshComponent.prototype, "materialColor", {
        get: /**
         * @return {?}
         */
        function () {
            return this._materialColor;
        },
        set: /**
         * @param {?} materialColor
         * @return {?}
         */
        function (materialColor) {
            // console.log('change color', materialColor);
            this._materialColor = materialColor;
            this.startLoading();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextMeshComponent.prototype, "text", {
        get: /**
         * @return {?}
         */
        function () {
            return this._text;
        },
        set: /**
         * @param {?} text
         * @return {?}
         */
        function (text) {
            this._text = text;
            this.startLoading();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TextMeshComponent.prototype.getMaterial = /**
     * @return {?}
     */
    function () {
        return appliedMaterial(this.materialColor, this.material, this.depthWrite);
    };
    /**
     * @protected
     * @return {?}
     */
    TextMeshComponent.prototype.loadLazyObject = /**
     * @protected
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                // console.log('TextMeshComponent.loadLazyObject');
                return [2 /*return*/, new Promise((/**
                     * @param {?} resolve
                     * @return {?}
                     */
                    function (resolve) {
                        /** @type {?} */
                        var loader = new THREE.FontLoader();
                        loader.load(_this.fontUrl, (/**
                         * @param {?} font
                         * @return {?}
                         */
                        function (font) {
                            /** @type {?} */
                            var geometry = new THREE.TextGeometry(_this.text, {
                                font: font,
                                size: _this.size,
                                height: _this.height,
                                curveSegments: _this.curveSegments,
                                bevelEnabled: _this.bevelEnabled,
                                bevelThickness: _this.bevelThickness,
                                bevelSize: _this.bevelSize,
                                bevelOffset: _this.bevelOffset,
                                bevelSegments: _this.bevelOffset
                            });
                            /** @type {?} */
                            var material = _this.getMaterial();
                            /** @type {?} */
                            var mesh = new THREE.Mesh(geometry, material);
                            mesh.castShadow = _this.castShadow;
                            mesh.receiveShadow = _this.receiveShadow;
                            if (_this.centered) {
                                fixCenter(mesh);
                            }
                            resolve(mesh);
                        }));
                    }))];
            });
        });
    };
    TextMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-text-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TextMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    TextMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    TextMeshComponent.propDecorators = {
        material: [{ type: Input }],
        materialColor: [{ type: Input }],
        text: [{ type: Input }],
        size: [{ type: Input }],
        height: [{ type: Input }],
        curveSegments: [{ type: Input }],
        bevelEnabled: [{ type: Input }],
        bevelThickness: [{ type: Input }],
        bevelSize: [{ type: Input }],
        bevelOffset: [{ type: Input }],
        bevelSegments: [{ type: Input }],
        fontUrl: [{ type: Input }],
        castShadow: [{ type: Input }],
        receiveShadow: [{ type: Input }],
        depthWrite: [{ type: Input }],
        centered: [{ type: Input }]
    };
    return TextMeshComponent;
}(AbstractLazyObject3D));
export { TextMeshComponent };
if (false) {
    /** @type {?} */
    TextMeshComponent.prototype.material;
    /**
     * @type {?}
     * @private
     */
    TextMeshComponent.prototype._materialColor;
    /**
     * @type {?}
     * @private
     */
    TextMeshComponent.prototype._text;
    /** @type {?} */
    TextMeshComponent.prototype.size;
    /** @type {?} */
    TextMeshComponent.prototype.height;
    /** @type {?} */
    TextMeshComponent.prototype.curveSegments;
    /** @type {?} */
    TextMeshComponent.prototype.bevelEnabled;
    /** @type {?} */
    TextMeshComponent.prototype.bevelThickness;
    /** @type {?} */
    TextMeshComponent.prototype.bevelSize;
    /** @type {?} */
    TextMeshComponent.prototype.bevelOffset;
    /** @type {?} */
    TextMeshComponent.prototype.bevelSegments;
    /** @type {?} */
    TextMeshComponent.prototype.fontUrl;
    /** @type {?} */
    TextMeshComponent.prototype.castShadow;
    /** @type {?} */
    TextMeshComponent.prototype.receiveShadow;
    /** @type {?} */
    TextMeshComponent.prototype.depthWrite;
    /** @type {?} */
    TextMeshComponent.prototype.centered;
    /**
     * @type {?}
     * @protected
     */
    TextMeshComponent.prototype.rendererService;
}
