// atft/esm2015/lib/camera/perspective-camera.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractCamera } from './abstract-camera';
import * as THREE from 'three';
import { RendererService } from '../renderer/renderer.service';
export class PerspectiveCameraComponent extends AbstractCamera {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
    }
    /**
     * @protected
     * @return {?}
     */
    createCamera() {
        // console.log('PerspectiveCameraComponent.createCamera');
        // let aspectRatio = undefined; // Updated later
        this.camera = new THREE.PerspectiveCamera(this.fov, undefined, this.near, this.far);
    }
    /**
     * @param {?} aspect
     * @return {?}
     */
    updateAspectRatio(aspect) {
        // console.log('PerspectiveCameraComponent.updateAspectRatio: ' + aspect);
        this.camera.aspect = aspect;
        this.camera.updateProjectionMatrix();
    }
}
PerspectiveCameraComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-perspective-camera',
                providers: [{ provide: AbstractCamera, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => PerspectiveCameraComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
PerspectiveCameraComponent.ctorParameters = () => [
    { type: RendererService }
];
PerspectiveCameraComponent.propDecorators = {
    fov: [{ type: Input }],
    near: [{ type: Input }],
    far: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PerspectiveCameraComponent.prototype.fov;
    /** @type {?} */
    PerspectiveCameraComponent.prototype.near;
    /** @type {?} */
    PerspectiveCameraComponent.prototype.far;
    /**
     * @type {?}
     * @protected
     */
    PerspectiveCameraComponent.prototype.rendererService;
}
