// atft/esm5/lib/object/helper/axes-helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var AxesHelperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AxesHelperComponent, _super);
    function AxesHelperComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AxesHelperComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        // console.log('AxesHelperComponent.newObject3DInstance');
        return new THREE.AxesHelper(this.size);
    };
    AxesHelperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-axes-helper',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return AxesHelperComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    AxesHelperComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    AxesHelperComponent.propDecorators = {
        size: [{ type: Input }]
    };
    return AxesHelperComponent;
}(AbstractObject3D));
export { AxesHelperComponent };
if (false) {
    /** @type {?} */
    AxesHelperComponent.prototype.size;
    /**
     * @type {?}
     * @protected
     */
    AxesHelperComponent.prototype.rendererService;
}
