// atft/esm5/lib/object/helper/grid-helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var GridHelperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(GridHelperComponent, _super);
    function GridHelperComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    GridHelperComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        // console.log('GridHelperComponent.newObject3DInstance');
        return new THREE.GridHelper(this.size, this.divisions);
    };
    GridHelperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-grid-helper',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return GridHelperComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    GridHelperComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    GridHelperComponent.propDecorators = {
        size: [{ type: Input }],
        divisions: [{ type: Input }]
    };
    return GridHelperComponent;
}(AbstractObject3D));
export { GridHelperComponent };
if (false) {
    /** @type {?} */
    GridHelperComponent.prototype.size;
    /** @type {?} */
    GridHelperComponent.prototype.divisions;
    /**
     * @type {?}
     * @protected
     */
    GridHelperComponent.prototype.rendererService;
}
