// atft/esm5/lib/object/helper/stats.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * @record
 */
export function StatsResults() { }
if (false) {
    /** @type {?} */
    StatsResults.prototype.ms;
    /** @type {?} */
    StatsResults.prototype.count;
    /** @type {?} */
    StatsResults.prototype.callsPerSecond;
}
var StatsService = /** @class */ (function () {
    function StatsService() {
        this.ms = 0;
        this.count = 0;
    }
    /**
     * @return {?}
     */
    StatsService.prototype.start = /**
     * @return {?}
     */
    function () {
        this.lastTime = Date.now();
    };
    /**
     * @return {?}
     */
    StatsService.prototype.end = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var deltaMs = Date.now() - this.lastTime;
        this.ms += deltaMs;
        this.count++;
    };
    /**
     * @return {?}
     */
    StatsService.prototype.getStats = /**
     * @return {?}
     */
    function () {
        return {
            ms: this.ms,
            count: this.count,
            callsPerSecond: Math.round(this.count / (this.ms / 1000))
        };
    };
    /**
     * @return {?}
     */
    StatsService.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.ms = 0;
        this.count = 0;
    };
    StatsService.decorators = [
        { type: Injectable }
    ];
    return StatsService;
}());
export { StatsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.lastTime;
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.ms;
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.count;
}
