// atft/esm5/lib/raycaster/raycaster-group.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Output } from '@angular/core';
import { AbstractObject3D } from '../object/abstract-object-3d';
import { RaycasterService } from './raycaster.service';
/**
 * Only components marked as atft-raycaster-group emit raycaster events.
 * NOTE: All childs are also considered by raycaster (very usefull, for composite components).
 */
var RaycasterGroupDirective = /** @class */ (function () {
    function RaycasterGroupDirective(host, raycasterService) {
        this.host = host;
        this.raycasterService = raycasterService;
        this.mouseEnter = new EventEmitter();
        this.mouseExit = new EventEmitter();
        this.mouseDown = new EventEmitter();
        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseExit = this.onMouseExit.bind(this);
        this.onMouseDown = this.onMouseDown.bind(this);
    }
    /**
     * @return {?}
     */
    RaycasterGroupDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.raycasterService.addGroup(this.host);
        this.subscribeEvents();
    };
    /**
     * @private
     * @return {?}
     */
    RaycasterGroupDirective.prototype.subscribeEvents = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = this.host.getObject();
        obj.addEventListener('mouseEnter', this.onMouseEnter);
        obj.addEventListener('mouseExit', this.onMouseExit);
        obj.addEventListener('mouseDown', this.onMouseDown);
    };
    /**
     * @private
     * @return {?}
     */
    RaycasterGroupDirective.prototype.unSubscribeEvents = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = this.host.getObject();
        obj.removeEventListener('mouseEnter', this.onMouseEnter);
        obj.removeEventListener('mouseExit', this.onMouseExit);
        obj.removeEventListener('mouseDown', this.onMouseDown);
    };
    /**
     * @private
     * @return {?}
     */
    RaycasterGroupDirective.prototype.onMouseExit = /**
     * @private
     * @return {?}
     */
    function () {
        this.mouseExit.emit();
    };
    /**
     * @private
     * @return {?}
     */
    RaycasterGroupDirective.prototype.onMouseEnter = /**
     * @private
     * @return {?}
     */
    function () {
        this.mouseEnter.emit();
    };
    /**
     * @private
     * @return {?}
     */
    RaycasterGroupDirective.prototype.onMouseDown = /**
     * @private
     * @return {?}
     */
    function () {
        this.mouseDown.emit();
    };
    /**
     * @return {?}
     */
    RaycasterGroupDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unSubscribeEvents();
    };
    RaycasterGroupDirective.decorators = [
        { type: Directive, args: [{ selector: '[atft-raycaster-group]' },] }
    ];
    /** @nocollapse */
    RaycasterGroupDirective.ctorParameters = function () { return [
        { type: AbstractObject3D },
        { type: RaycasterService }
    ]; };
    RaycasterGroupDirective.propDecorators = {
        mouseEnter: [{ type: Output }],
        mouseExit: [{ type: Output }],
        mouseDown: [{ type: Output }]
    };
    return RaycasterGroupDirective;
}());
export { RaycasterGroupDirective };
if (false) {
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseEnter;
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseExit;
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseDown;
    /**
     * @type {?}
     * @private
     */
    RaycasterGroupDirective.prototype.host;
    /**
     * @type {?}
     * @private
     */
    RaycasterGroupDirective.prototype.raycasterService;
}
