// atft/esm2015/lib/object/loader/obj-loader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractModelLoader } from './abstract-model-loader';
import { OBJLoader } from 'three/examples/jsm/loaders/OBJLoader';
import { MTLLoader } from 'three/examples/jsm/loaders/MTLLoader';
import { RendererService } from '../../renderer/renderer.service';
/**
 * Component for employing THREE.OBJLoader to load [Wavefront *.obj files][1].
 *
 * [1]: https://en.wikipedia.org/wiki/Wavefront_.obj_file
 */
export class ObjLoaderComponent extends AbstractModelLoader {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.loader = new OBJLoader();
        this.mtlLoader = new MTLLoader();
    }
    /**
     * Path relative to which resources and textures within the loaded obj file are loaded.
     * @deprecated Rather use resourcePath
     * @param {?} newTexturePath
     * @return {?}
     */
    set texturePath(newTexturePath) {
        this.resourcePath = newTexturePath;
    }
    /**
     * @protected
     * @return {?}
     */
    loadLazyObject() {
        return tslib_1.__awaiter(this, void 0, void 0, /** @this {!ObjLoaderComponent} */ function* () {
            // Preloading step for the material
            /** @type {?} */
            const preloadingStep = new Promise((/**
             * @param {?} resolve
             * @param {?} reject
             * @return {?}
             */
            (resolve, reject) => {
                if (this.material === undefined) {
                    // No preloading necessary
                    resolve();
                }
                else {
                    // TODO: If typings of mtlLoader are included in the Three.js NPM
                    // package, remove this 'any' cast.
                    ((/** @type {?} */ (this.mtlLoader))).setResourcePath(this.resourcePath);
                    this.mtlLoader.load(this.material, (/**
                     * @param {?} materialCreator
                     * @return {?}
                     */
                    materialCreator => {
                        materialCreator.preload();
                        this.loader.setMaterials(materialCreator);
                        resolve();
                    }), undefined, reject);
                }
            }));
            // Await preloading and load final model
            return preloadingStep.then((/**
             * @return {?}
             */
            () => {
                return new Promise((/**
                 * @param {?} resolve
                 * @param {?} reject
                 * @return {?}
                 */
                (resolve, reject) => {
                    this.loader.load(this.model, (/**
                     * @param {?} model
                     * @return {?}
                     */
                    model => {
                        resolve(model);
                    }), undefined, reject);
                }));
            }));
        });
    }
}
ObjLoaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-obj-loader',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ObjLoaderComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
ObjLoaderComponent.ctorParameters = () => [
    { type: RendererService }
];
ObjLoaderComponent.propDecorators = {
    material: [{ type: Input }],
    texturePath: [{ type: Input }],
    resourcePath: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ObjLoaderComponent.prototype.loader;
    /**
     * @type {?}
     * @private
     */
    ObjLoaderComponent.prototype.mtlLoader;
    /** @type {?} */
    ObjLoaderComponent.prototype.material;
    /**
     * Path relative to which resources and textures within the loaded obj file are loaded.
     * @type {?}
     */
    ObjLoaderComponent.prototype.resourcePath;
    /**
     * @type {?}
     * @protected
     */
    ObjLoaderComponent.prototype.rendererService;
}
