// atft/esm2015/lib/object/loader/object-loader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractModelLoader } from './abstract-model-loader';
import * as THREE from 'three';
import { RendererService } from '../../renderer/renderer.service';
export class ObjectLoaderComponent extends AbstractModelLoader {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.loader = new THREE.ObjectLoader();
    }
    /**
     * @protected
     * @return {?}
     */
    loadLazyObject() {
        return tslib_1.__awaiter(this, void 0, void 0, /** @this {!ObjectLoaderComponent} */ function* () {
            // console.log('ObjectLoaderComponent.loadLazyObject');
            return new Promise((/**
             * @param {?} resolve
             * @param {?} reject
             * @return {?}
             */
            (resolve, reject) => {
                this.loader.load(this.model, (/**
                 * @param {?} model
                 * @return {?}
                 */
                model => {
                    // BUG #95: it seems that some textures loaded after last render (and model has black texture instead)
                    resolve(model);
                }), undefined, reject);
            }));
        });
    }
}
ObjectLoaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-object-loader',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ObjectLoaderComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
ObjectLoaderComponent.ctorParameters = () => [
    { type: RendererService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ObjectLoaderComponent.prototype.loader;
    /**
     * @type {?}
     * @protected
     */
    ObjectLoaderComponent.prototype.rendererService;
}
