// atft/esm5/lib/object/light/directional-light.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { appliedColor } from '../../util/applied-color';
import { RendererService } from '../../renderer/renderer.service';
var DirectionalLightComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DirectionalLightComponent, _super);
    function DirectionalLightComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.color = 0xffffff;
        _this.intensity = 1;
        // by default, target is 0,0,0
        _this.target = new THREE.Object3D();
        _this.castShadow = true;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    DirectionalLightComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var light = new THREE.DirectionalLight(appliedColor(this.color), this.intensity);
        light.target = this.target;
        if (this.castShadow === true) {
            light.castShadow = this.castShadow;
            // TODO: props
            light.shadow.camera.top = 100;
            light.shadow.camera.bottom = -100;
            light.shadow.camera.left = -100;
            light.shadow.camera.right = 100;
            light.shadow.camera.near = 0.1;
            light.shadow.camera.far = 500;
            light.shadow.mapSize.set(1024, 1024);
            light.shadow.bias = -0.001;
        }
        return light;
    };
    DirectionalLightComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-directional-light',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return DirectionalLightComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    DirectionalLightComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    DirectionalLightComponent.propDecorators = {
        color: [{ type: Input }],
        intensity: [{ type: Input }],
        target: [{ type: Input }],
        castShadow: [{ type: Input }]
    };
    return DirectionalLightComponent;
}(AbstractObject3D));
export { DirectionalLightComponent };
if (false) {
    /** @type {?} */
    DirectionalLightComponent.prototype.color;
    /** @type {?} */
    DirectionalLightComponent.prototype.intensity;
    /** @type {?} */
    DirectionalLightComponent.prototype.target;
    /** @type {?} */
    DirectionalLightComponent.prototype.castShadow;
    /**
     * @type {?}
     * @protected
     */
    DirectionalLightComponent.prototype.rendererService;
}
