// atft/esm5/lib/object/light/hemisphere-light.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { appliedColor } from '../../util/applied-color';
import { RendererService } from '../../renderer/renderer.service';
var HemisphereLightComponent = /** @class */ (function (_super) {
    tslib_1.__extends(HemisphereLightComponent, _super);
    function HemisphereLightComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.skyColor = 0xffffff;
        _this.groundColor = 0x444444;
        _this.intensity = 1;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    HemisphereLightComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var light = new THREE.HemisphereLight(appliedColor(this.skyColor), appliedColor(this.groundColor), appliedColor(this.intensity));
        return light;
    };
    HemisphereLightComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-hemisphere-light',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return HemisphereLightComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    HemisphereLightComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    HemisphereLightComponent.propDecorators = {
        skyColor: [{ type: Input }],
        groundColor: [{ type: Input }],
        intensity: [{ type: Input }]
    };
    return HemisphereLightComponent;
}(AbstractObject3D));
export { HemisphereLightComponent };
if (false) {
    /** @type {?} */
    HemisphereLightComponent.prototype.skyColor;
    /** @type {?} */
    HemisphereLightComponent.prototype.groundColor;
    /** @type {?} */
    HemisphereLightComponent.prototype.intensity;
    /**
     * @type {?}
     * @protected
     */
    HemisphereLightComponent.prototype.rendererService;
}
