// atft/esm2015/lib/raycaster/raycaster-group.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Output } from '@angular/core';
import { AbstractObject3D } from '../object/abstract-object-3d';
import { RaycasterService } from './raycaster.service';
/**
 * Only components marked as atft-raycaster-group emit raycaster events.
 * NOTE: All childs are also considered by raycaster (very usefull, for composite components).
 */
export class RaycasterGroupDirective {
    /**
     * @param {?} host
     * @param {?} raycasterService
     */
    constructor(host, raycasterService) {
        this.host = host;
        this.raycasterService = raycasterService;
        this.mouseEnter = new EventEmitter();
        this.mouseExit = new EventEmitter();
        this.mouseDown = new EventEmitter();
        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseExit = this.onMouseExit.bind(this);
        this.onMouseDown = this.onMouseDown.bind(this);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.raycasterService.addGroup(this.host);
        this.subscribeEvents();
    }
    /**
     * @private
     * @return {?}
     */
    subscribeEvents() {
        /** @type {?} */
        const obj = this.host.getObject();
        obj.addEventListener('mouseEnter', this.onMouseEnter);
        obj.addEventListener('mouseExit', this.onMouseExit);
        obj.addEventListener('mouseDown', this.onMouseDown);
    }
    /**
     * @private
     * @return {?}
     */
    unSubscribeEvents() {
        /** @type {?} */
        const obj = this.host.getObject();
        obj.removeEventListener('mouseEnter', this.onMouseEnter);
        obj.removeEventListener('mouseExit', this.onMouseExit);
        obj.removeEventListener('mouseDown', this.onMouseDown);
    }
    /**
     * @private
     * @return {?}
     */
    onMouseExit() {
        this.mouseExit.emit();
    }
    /**
     * @private
     * @return {?}
     */
    onMouseEnter() {
        this.mouseEnter.emit();
    }
    /**
     * @private
     * @return {?}
     */
    onMouseDown() {
        this.mouseDown.emit();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unSubscribeEvents();
    }
}
RaycasterGroupDirective.decorators = [
    { type: Directive, args: [{ selector: '[atft-raycaster-group]' },] }
];
/** @nocollapse */
RaycasterGroupDirective.ctorParameters = () => [
    { type: AbstractObject3D },
    { type: RaycasterService }
];
RaycasterGroupDirective.propDecorators = {
    mouseEnter: [{ type: Output }],
    mouseExit: [{ type: Output }],
    mouseDown: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseEnter;
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseExit;
    /** @type {?} */
    RaycasterGroupDirective.prototype.mouseDown;
    /**
     * @type {?}
     * @private
     */
    RaycasterGroupDirective.prototype.host;
    /**
     * @type {?}
     * @private
     */
    RaycasterGroupDirective.prototype.raycasterService;
}
