// atft/esm2015/lib/raycaster/raycaster.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as THREE from 'three';
export class RaycasterService {
    constructor() {
        this.raycaster = new THREE.Raycaster();
        this.enabled = false;
        this.groups = [];
        this.paused = false;
        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseDown = this.onMouseDown.bind(this);
        this.onTouchStart = this.onTouchStart.bind(this);
        this.subscribe();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.disable();
        this.unsubscribe();
    }
    /**
     * @private
     * @return {?}
     */
    subscribe() {
        window.addEventListener('mousemove', this.onMouseMove);
        window.addEventListener('mousedown', this.onMouseDown);
        window.addEventListener('touchstart', this.onTouchStart);
    }
    /**
     * @private
     * @return {?}
     */
    unsubscribe() {
        // console.log('unsubscribe raycaster');
        window.removeEventListener('mousemove', this.onMouseMove);
        window.removeEventListener('mousedown', this.onMouseDown);
        window.removeEventListener('touchstart', this.onTouchStart);
    }
    /**
     * @return {?}
     */
    enable() {
        this.enabled = true;
    }
    /**
     * @return {?}
     */
    disable() {
        this.enabled = false;
    }
    /**
     * @return {?}
     */
    pause() {
        this.paused = true;
    }
    /**
     * @return {?}
     */
    resume() {
        this.paused = false;
    }
    /**
     * @return {?}
     */
    get isEnabled() {
        return this.enabled;
    }
    /**
     * @param {?} camera
     * @return {?}
     */
    setCamera(camera) {
        // console.log('Add camera to raycaster', camera);
        this.camera = camera;
    }
    /**
     * @param {?} group
     * @return {?}
     */
    addGroup(group) {
        // console.log('RaycasterService.addGroup', group.name, group);
        this.groups.push(group);
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onMouseMove(event) {
        if (!this.isReady()) {
            return;
        }
        event.preventDefault();
        /** @type {?} */
        const i = this.getFirstIntersectedGroup(event.layerX, event.layerY);
        if (!this.selected || this.selected !== i) {
            if (this.selected) {
                this.selected.dispatchEvent({ type: 'mouseExit' });
                this.selected = null;
            }
            if (i) {
                this.selected = i;
                // console.log('RaycasterService.mouseEnter', i);
                this.selected.dispatchEvent({ type: 'mouseEnter' });
            }
        }
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onMouseDown(event) {
        if (!this.isReady(true)) {
            return;
        }
        event.preventDefault();
        /** @type {?} */
        const i = this.getFirstIntersectedGroup(event.layerX, event.layerY);
        if (i) {
            i.dispatchEvent({ type: 'mouseDown' });
        }
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onTouchStart(event) {
        // console.log(event);
        if (!this.isReady()) {
            return;
        }
        event.preventDefault();
        /** @type {?} */
        const i = this.getFirstIntersectedGroup(event.touches[0].clientX, event.touches[0].clientY);
        if (i) {
            i.dispatchEvent({ type: 'mouseDown' });
        }
    }
    /**
     * @private
     * @param {?=} ignorePaused
     * @return {?}
     */
    isReady(ignorePaused) {
        return this.enabled
            && (ignorePaused || !this.paused)
            && this.camera
            && this.camera.camera
            && this.groups
            && this.groups.length > 0;
    }
    /**
     * @private
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    getFirstIntersectedGroup(x, y) {
        x = (x / window.innerWidth) * 2 - 1;
        y = -(y / window.innerHeight) * 2 + 1;
        /** @type {?} */
        const mouseVector = new THREE.Vector3(x, y, 0.5);
        this.raycaster.setFromCamera(mouseVector, this.camera.camera);
        // loop across all groups. Try to find the group with nearest distance.
        /** @type {?} */
        let nearestIntersection;
        /** @type {?} */
        let nearestGroup;
        for (let k = 0; k < this.groups.length; k++) {
            /** @type {?} */
            const i = this.groups[k].getObject();
            /** @type {?} */
            const intersection = this.raycaster.intersectObject(i, true);
            if (intersection.length > 0 && (!nearestIntersection || nearestIntersection.distance > intersection[0].distance)) {
                nearestIntersection = intersection[0];
                nearestGroup = i;
            }
        }
        // return the group with nearest distance
        if (nearestGroup) {
            return nearestGroup;
        }
        else {
            return;
        }
    }
}
RaycasterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RaycasterService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.raycaster;
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.selected;
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.enabled;
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.camera;
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.groups;
    /**
     * @type {?}
     * @private
     */
    RaycasterService.prototype.paused;
}
