// atft/esm2015/lib/object/connector/abstract-connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
/**
 * @abstract
 * @template T
 */
export class AbstractConnector extends AbstractObject3D {
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        /** @type {?} */
        const mesh = this.createConnectorObject();
        this.watchObjects();
        return mesh;
    }
    /**
     * @private
     * @return {?}
     */
    watchObjects() {
        this.source.changed.subscribe((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            this.updateLineGeometry();
        }));
        this.target.changed.subscribe((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            this.updateLineGeometry();
        }));
    }
    /**
     * @protected
     * @return {?}
     */
    getLineGeometry() {
        /** @type {?} */
        const geo = new THREE.Geometry();
        if (!this.source || !this.target) {
            throw new Error('AbstractConnector: source or target inputs are missing!');
        }
        geo.vertices.push(this.source.getObject().position);
        geo.vertices.push(this.target.getObject().position);
        return geo;
    }
}
AbstractConnector.propDecorators = {
    source: [{ type: Input }],
    target: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AbstractConnector.prototype.source;
    /** @type {?} */
    AbstractConnector.prototype.target;
    /**
     * Create line mesh
     * @abstract
     * @return {?}
     */
    AbstractConnector.prototype.createConnectorObject = function () { };
    /**
     * If at least one line end (source or target object)  changed, then line geoetry should be updated as well
     * // TODO: Calculate only when source/target positions were changed
     * @abstract
     * @return {?}
     */
    AbstractConnector.prototype.updateLineGeometry = function () { };
}
