// atft/esm2015/lib/object/connector/line-connector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractConnector } from './abstract-connector';
import { appliedColor } from '../../util/applied-color';
import { RendererService } from '../../renderer/renderer.service';
export class LineConnectorComponent extends AbstractConnector {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.materialColor = 0xffff00;
    }
    /**
     * @return {?}
     */
    createConnectorObject() {
        this.geometry = this.getLineGeometry();
        /** @type {?} */
        const material = new THREE.LineBasicMaterial({
            color: appliedColor(this.materialColor),
            /**
             * NOTE: linewidth=1: Due to limitations of the OpenGL Core Profile
             * with the WebGL renderer on most platforms linewidth will always be 1
             * regardless of the set value.
             */
            linewidth: 1
        });
        /** @type {?} */
        const line = new THREE.Line(this.geometry, material);
        return line;
    }
    /**
     * @return {?}
     */
    updateLineGeometry() {
        this.geometry.verticesNeedUpdate = true;
        this.rendererService.render();
    }
}
LineConnectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-line-connector',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => LineConnectorComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
LineConnectorComponent.ctorParameters = () => [
    { type: RendererService }
];
LineConnectorComponent.propDecorators = {
    materialColor: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    LineConnectorComponent.prototype.materialColor;
    /**
     * @type {?}
     * @private
     */
    LineConnectorComponent.prototype.geometry;
    /**
     * @type {?}
     * @protected
     */
    LineConnectorComponent.prototype.rendererService;
}
