// atft/esm5/lib/control/abstract-orbit-controls.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChild, ContentChildren, QueryList } from '@angular/core';
import { AbstractCamera } from '../camera/abstract-camera';
import { RendererCanvasComponent } from '../renderer/renderer-canvas.component';
/**
 * @abstract
 * @template T
 */
var AbstractOrbitControls = /** @class */ (function () {
    function AbstractOrbitControls(rendererService, raycasterService) {
        this.rendererService = rendererService;
        this.raycasterService = raycasterService;
        // console.log('OrbitControlsComponent.constructor');
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    AbstractOrbitControls.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        // If the THREE.js OrbitControls are not set up yet, we do not need to update
        // anything as they will pick the new values from the @Input properties automatically
        // upon creation.
        if (!this.controls) {
            return;
        }
        if (changes['listeningControlElement']) {
            // The DOM element the OrbitControls listen on cannot be changed once an
            // OrbitControls object is created. We thus need to recreate it.
            this.controls.dispose();
            this.setUpControls();
        }
    };
    /**
     * @return {?}
     */
    AbstractOrbitControls.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.controls) {
            this.controls.dispose();
        }
    };
    /**
     * @private
     * @return {?}
     */
    AbstractOrbitControls.prototype.configureListeners = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.controls.addEventListener('change', (/**
         * @return {?}
         */
        function () {
            _this.rendererService.render();
        }));
        // don't raycast during rotation/damping/panning
        if (this.raycasterService.isEnabled) {
            this.controls.addEventListener('start', (/**
             * @return {?}
             */
            function () {
                _this.raycasterService.pause();
            }));
            this.controls.addEventListener('end', (/**
             * @return {?}
             */
            function () {
                _this.raycasterService.resume();
            }));
        }
    };
    /**
     * @return {?}
     */
    AbstractOrbitControls.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // console.log('OrbitControlsComponent.ngAfterViewInit');
        if (this.childCameras === undefined || this.childCameras.first === undefined) {
            throw new Error('Camera is not found');
        }
        if (!this.webGlRenderer) {
            throw new Error('webGlRenderer is not found');
        }
        this.listeningControlElement = this.webGlRenderer.renderPane;
        this.setUpControls();
        this.configureListeners();
        this.rendererService.render();
    };
    AbstractOrbitControls.propDecorators = {
        childCameras: [{ type: ContentChildren, args: [AbstractCamera, { descendants: true },] }],
        webGlRenderer: [{ type: ContentChild, args: [RendererCanvasComponent, { static: false },] }]
    };
    return AbstractOrbitControls;
}());
export { AbstractOrbitControls };
if (false) {
    /** @type {?} */
    AbstractOrbitControls.prototype.childCameras;
    /** @type {?} */
    AbstractOrbitControls.prototype.webGlRenderer;
    /**
     * The element on whose native element the orbit control will listen for mouse events.
     *
     * Note that keyboard events are still listened for on the global window object, this is
     * a known issue from Three.js: https://github.com/mrdoob/three.js/pull/10315
     * @type {?}
     * @protected
     */
    AbstractOrbitControls.prototype.listeningControlElement;
    /**
     * @type {?}
     * @protected
     */
    AbstractOrbitControls.prototype.controls;
    /**
     * @type {?}
     * @protected
     */
    AbstractOrbitControls.prototype.rendererService;
    /**
     * @type {?}
     * @protected
     */
    AbstractOrbitControls.prototype.raycasterService;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    AbstractOrbitControls.prototype.setUpControls = function () { };
}
