// atft/esm2015/lib/object/mesh/abstract-css3d-mesh.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { calculateSize } from '../../util';
/**
 * @abstract
 */
export class AbstractCss3dMesh extends AbstractMesh {
    constructor() {
        super(...arguments);
        this.width = 1.0;
        this.height = 1.0;
        // 1280 / 720 = 1.777777778
        this.resolutionX = 1280;
        this.resolutionY = 720;
        this.widthSegments = 1;
        this.heightSegments = 1;
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        /** @type {?} */
        const geometry = new THREE.PlaneBufferGeometry(this.width, this.height, this.widthSegments, this.heightSegments);
        /** @type {?} */
        const material = this.getMaterial();
        /** @type {?} */
        const mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        // Create and scale css3d object:
        /** @type {?} */
        const meshSize = calculateSize(mesh);
        /** @type {?} */
        const css3dObject = this.createCss3dObject();
        this.scaleCss3d(css3dObject, meshSize);
        // append to mesh:
        mesh.add(css3dObject);
        return mesh;
    }
    /**
     * @protected
     * @param {?} group
     * @param {?} max
     * @return {?}
     */
    scaleCss3d(group, max) {
        /** @type {?} */
        const box = new THREE.Vector3(this.resolutionX, this.resolutionY, 0);
        /** @type {?} */
        const scaleX = max.x / box.x;
        /** @type {?} */
        const scaleY = max.y / box.y;
        /** @type {?} */
        const scaleZ = max.z / box.z;
        group.scale.set((scaleX < 1 ? scaleX : 1), (scaleY < 1 ? scaleY : 1), (scaleZ < 1 ? scaleZ : 1));
    }
    /**
     * @protected
     * @return {?}
     */
    getMaterial() {
        // TODO: Renderer clear color
        /** @type {?} */
        const transparentMaterial = new THREE.MeshBasicMaterial({
            color: 0x000000,
            opacity: 0,
            side: THREE.DoubleSide
        });
        return transparentMaterial;
    }
}
AbstractCss3dMesh.propDecorators = {
    width: [{ type: Input }],
    height: [{ type: Input }],
    resolutionX: [{ type: Input }],
    resolutionY: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AbstractCss3dMesh.prototype.width;
    /** @type {?} */
    AbstractCss3dMesh.prototype.height;
    /** @type {?} */
    AbstractCss3dMesh.prototype.resolutionX;
    /** @type {?} */
    AbstractCss3dMesh.prototype.resolutionY;
    /**
     * @type {?}
     * @private
     */
    AbstractCss3dMesh.prototype.widthSegments;
    /**
     * @type {?}
     * @private
     */
    AbstractCss3dMesh.prototype.heightSegments;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    AbstractCss3dMesh.prototype.createCss3dObject = function () { };
}
