// atft/esm2015/lib/object/mesh/abstract-mesh-3d.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { appliedMaterial } from '../../util';
/**
 * @abstract
 */
export class AbstractMesh extends AbstractObject3D {
    constructor() {
        super(...arguments);
        this.materialColor = 0x5DADE2;
        this.castShadow = true;
        this.receiveShadow = true;
        this.depthWrite = true;
    }
    /**
     * @protected
     * @return {?}
     */
    getMaterial() {
        return appliedMaterial(this.materialColor, this.material, this.depthWrite);
    }
    /**
     * @protected
     * @param {?} mesh
     * @return {?}
     */
    applyShadowProps(mesh) {
        mesh.castShadow = this.castShadow;
        mesh.receiveShadow = this.receiveShadow;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        if (!this.getObject()) {
            return;
        }
        /** @type {?} */
        let mustRerender = false;
        if (['material', 'materialColor', 'depthWrite'].some((/**
         * @param {?} propName
         * @return {?}
         */
        propName => propName in changes))) {
            this.applyMaterial();
            mustRerender = true;
        }
        if (mustRerender) {
            this.rendererService.render();
        }
    }
    /**
     * @return {?}
     */
    applyMaterial() {
        this.getObject().material = this.getMaterial();
    }
}
AbstractMesh.propDecorators = {
    material: [{ type: Input }],
    materialColor: [{ type: Input }],
    castShadow: [{ type: Input }],
    receiveShadow: [{ type: Input }],
    depthWrite: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AbstractMesh.prototype.material;
    /** @type {?} */
    AbstractMesh.prototype.materialColor;
    /** @type {?} */
    AbstractMesh.prototype.castShadow;
    /** @type {?} */
    AbstractMesh.prototype.receiveShadow;
    /** @type {?} */
    AbstractMesh.prototype.depthWrite;
}
