// atft/esm2015/lib/object/mesh/css3d-video-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { CSS3DObject } from 'three/examples/jsm/renderers/CSS3DRenderer';
import { AbstractCss3dMesh } from './abstract-css3d-mesh';
import { RendererService } from '../../renderer/renderer.service';
export class Css3dVideoMeshComponent extends AbstractCss3dMesh {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.video = document.createElement('video');
    }
    /**
     * @protected
     * @return {?}
     */
    createCss3dObject() {
        // 1. Create DIV
        /** @type {?} */
        const div = document.createElement('div');
        div.style.width = this.resolutionX + 'px';
        div.style.height = this.resolutionY + 'px';
        div.style.backgroundColor = '#000000';
        div.id = 'myWrapper';
        // 2. Create Video Player with source
        this.video.autoplay = false;
        this.video.loop = true;
        this.video.muted = true;
        /** @type {?} */
        const source = document.createElement('source');
        source.src = this.videoSrc;
        source.type = 'video/mp4';
        this.video.appendChild(source);
        div.appendChild(this.video);
        // 3. Create CSS3DObject
        /** @type {?} */
        const object = new CSS3DObject(div);
        return object;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.video.load();
        this.video.addEventListener('canplay', (/**
         * @return {?}
         */
        () => {
            console.log('EVENT');
            this.video.play();
        }));
    }
}
Css3dVideoMeshComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-css3d-video-mesh',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => Css3dVideoMeshComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
Css3dVideoMeshComponent.ctorParameters = () => [
    { type: RendererService }
];
Css3dVideoMeshComponent.propDecorators = {
    videoSrc: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    Css3dVideoMeshComponent.prototype.videoSrc;
    /** @type {?} */
    Css3dVideoMeshComponent.prototype.video;
    /**
     * @type {?}
     * @protected
     */
    Css3dVideoMeshComponent.prototype.rendererService;
}
