// atft/esm2015/lib/object/mesh/cylinder-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
export class CylinderMeshComponent extends AbstractMesh {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.radiusTop = 1.0;
        this.radiusBottom = 1.0;
        this.height = 1.0;
        this.radialSegments = 8;
        this.heightSegments = 1;
        this.openEnded = false;
        this.thetaStart = 0.0;
        this.thetaLength = 2 * Math.PI;
        // console.log('CylinderMeshComponent.constructor');
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        // console.log('CylinderMeshComponent.newObject3DInstance');
        /** @type {?} */
        const geometry = new THREE.CylinderGeometry(this.radiusTop, this.radiusBottom, this.height, this.radialSegments, this.heightSegments, this.openEnded, this.thetaStart, this.thetaLength);
        /** @type {?} */
        const material = this.getMaterial();
        /** @type {?} */
        const mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        return mesh;
    }
}
CylinderMeshComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-cylinder-mesh',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => CylinderMeshComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
CylinderMeshComponent.ctorParameters = () => [
    { type: RendererService }
];
CylinderMeshComponent.propDecorators = {
    radiusTop: [{ type: Input }],
    radiusBottom: [{ type: Input }],
    height: [{ type: Input }],
    radialSegments: [{ type: Input }],
    heightSegments: [{ type: Input }],
    openEnded: [{ type: Input }],
    thetaStart: [{ type: Input }],
    thetaLength: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CylinderMeshComponent.prototype.radiusTop;
    /** @type {?} */
    CylinderMeshComponent.prototype.radiusBottom;
    /** @type {?} */
    CylinderMeshComponent.prototype.height;
    /** @type {?} */
    CylinderMeshComponent.prototype.radialSegments;
    /** @type {?} */
    CylinderMeshComponent.prototype.heightSegments;
    /** @type {?} */
    CylinderMeshComponent.prototype.openEnded;
    /** @type {?} */
    CylinderMeshComponent.prototype.thetaStart;
    /** @type {?} */
    CylinderMeshComponent.prototype.thetaLength;
    /**
     * @type {?}
     * @protected
     */
    CylinderMeshComponent.prototype.rendererService;
}
