// atft/esm5/lib/animation/animation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { RendererService } from '../renderer/renderer.service';
/**
 * Animation service emits animate event, which should be used by animated components for animation logic.
 * NOTE: this service is for the performance optimization: requestAnimationFrame and render is called once.
 */
var AnimationService = /** @class */ (function () {
    function AnimationService(rendererService) {
        this.rendererService = rendererService;
        /**
         * Subscribe for animation frame creation (change position and etc.)
         * Avoid render() execution, it's called only once when all components updated animation frame.
         */
        this.animate = new EventEmitter();
        this.enabled = false;
        this.animationStep = this.animationStep.bind(this);
    }
    /**
     * Start the animation loop.
     */
    /**
     * Start the animation loop.
     * @return {?}
     */
    AnimationService.prototype.start = /**
     * Start the animation loop.
     * @return {?}
     */
    function () {
        if (!this.enabled) {
            this.enabled = true;
            this.animationStep();
        }
    };
    /**
     * Stop all animations.
     */
    /**
     * Stop all animations.
     * @return {?}
     */
    AnimationService.prototype.stop = /**
     * Stop all animations.
     * @return {?}
     */
    function () {
        if (this.enabled) {
            this.enabled = false;
        }
    };
    /**
     * @private
     * @return {?}
     */
    AnimationService.prototype.animationStep = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.enabled) {
            requestAnimationFrame(this.animationStep);
            if (this.animate.observers.length > 0) {
                this.animate.emit();
                /**
                 * When all components updated animation, render event is emitted.
                 * Main renderer subscribed to this event emitter.
                 */
                this.rendererService.render();
            }
        }
    };
    AnimationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AnimationService.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    return AnimationService;
}());
export { AnimationService };
if (false) {
    /**
     * Subscribe for animation frame creation (change position and etc.)
     * Avoid render() execution, it's called only once when all components updated animation frame.
     * @type {?}
     */
    AnimationService.prototype.animate;
    /**
     * @type {?}
     * @private
     */
    AnimationService.prototype.enabled;
    /**
     * @type {?}
     * @private
     */
    AnimationService.prototype.rendererService;
}
