// atft/esm2015/lib/control/map-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { MapControls } from 'three/examples/jsm/controls/OrbitControls';
import { RendererService } from '../renderer/renderer.service';
import { AnimationService } from '../animation/animation.service';
import { RaycasterService } from '../raycaster/raycaster.service';
import { AbstractOrbitControls } from './abstract-orbit-controls';
export class MapControlsComponent extends AbstractOrbitControls {
    /**
     * @param {?} rendererService
     * @param {?} raycasterService
     * @param {?} animationService
     */
    constructor(rendererService, raycasterService, animationService) {
        super(rendererService, raycasterService);
        this.rendererService = rendererService;
        this.raycasterService = raycasterService;
        this.animationService = animationService;
        this.rotateSpeed = 1.0;
        this.zoomSpeed = 1.2;
        this.autoRotate = false;
        this.autoRotateSpeed = 0.5;
        this.enableDamping = false;
        this.dampingFactor = 0.1;
        this.screenSpacePanning = false;
        this.minDistance = 20;
        this.maxDistance = 200;
        this.maxPolarAngle = Math.PI / 2 - 0.1;
        this.panSpeed = 1.2;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this.controls) {
            return;
        }
        super.ngOnChanges(changes);
        if (changes['rotateSpeed']) {
            this.controls.rotateSpeed = this.rotateSpeed;
        }
        if (changes['zoomSpeed']) {
            this.controls.zoomSpeed = this.zoomSpeed;
        }
        // TODO: add others
    }
    /**
     * @protected
     * @return {?}
     */
    setUpControls() {
        this.controls = new MapControls(this.childCameras.first.camera, this.listeningControlElement && this.listeningControlElement.nativeElement);
        this.controls.rotateSpeed = this.rotateSpeed;
        this.controls.zoomSpeed = this.zoomSpeed;
        this.controls.panSpeed = this.panSpeed;
        this.controls.autoRotate = this.autoRotate;
        this.controls.autoRotateSpeed = this.autoRotateSpeed;
        this.controls.enableDamping = this.enableDamping; // an animation loop is required when either damping or auto-rotation are enabled
        this.controls.dampingFactor = this.dampingFactor;
        this.controls.screenSpacePanning = this.screenSpacePanning;
        this.controls.minDistance = this.minDistance;
        this.controls.maxDistance = this.maxDistance;
        this.controls.maxPolarAngle = this.maxPolarAngle;
        this.controls.update();
        // Advanced animation:
        if (this.autoRotate || this.enableDamping) {
            this.animationService.animate.subscribe((/**
             * @return {?}
             */
            () => {
                this.controls.update();
            }));
            this.controls.addEventListener('change', (/**
             * @return {?}
             */
            () => {
                this.rendererService.render();
            }));
            this.animationService.start();
        }
        this.rendererService.render();
    }
}
MapControlsComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-map-controls',
                template: `
      <ng-content></ng-content>`,
                styles: [":host{display:flex;flex:1}"]
            }] }
];
/** @nocollapse */
MapControlsComponent.ctorParameters = () => [
    { type: RendererService },
    { type: RaycasterService },
    { type: AnimationService }
];
MapControlsComponent.propDecorators = {
    rotateSpeed: [{ type: Input }],
    zoomSpeed: [{ type: Input }],
    autoRotate: [{ type: Input }],
    autoRotateSpeed: [{ type: Input }],
    enableDamping: [{ type: Input }],
    dampingFactor: [{ type: Input }],
    screenSpacePanning: [{ type: Input }],
    minDistance: [{ type: Input }],
    maxDistance: [{ type: Input }],
    maxPolarAngle: [{ type: Input }],
    panSpeed: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MapControlsComponent.prototype.rotateSpeed;
    /** @type {?} */
    MapControlsComponent.prototype.zoomSpeed;
    /** @type {?} */
    MapControlsComponent.prototype.autoRotate;
    /** @type {?} */
    MapControlsComponent.prototype.autoRotateSpeed;
    /** @type {?} */
    MapControlsComponent.prototype.enableDamping;
    /** @type {?} */
    MapControlsComponent.prototype.dampingFactor;
    /** @type {?} */
    MapControlsComponent.prototype.screenSpacePanning;
    /** @type {?} */
    MapControlsComponent.prototype.minDistance;
    /** @type {?} */
    MapControlsComponent.prototype.maxDistance;
    /** @type {?} */
    MapControlsComponent.prototype.maxPolarAngle;
    /** @type {?} */
    MapControlsComponent.prototype.panSpeed;
    /**
     * @type {?}
     * @protected
     */
    MapControlsComponent.prototype.rendererService;
    /**
     * @type {?}
     * @protected
     */
    MapControlsComponent.prototype.raycasterService;
    /**
     * @type {?}
     * @protected
     */
    MapControlsComponent.prototype.animationService;
}
