// atft/esm2015/lib/control/orbit-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls';
import { RendererService } from '../renderer/renderer.service';
import { RaycasterService } from '../raycaster/raycaster.service';
import { AbstractOrbitControls } from './abstract-orbit-controls';
export class OrbitControlsComponent extends AbstractOrbitControls {
    /**
     * @param {?} rendererService
     * @param {?} raycasterService
     */
    constructor(rendererService, raycasterService) {
        super(rendererService, raycasterService);
        this.rendererService = rendererService;
        this.raycasterService = raycasterService;
        this.rotateSpeed = 1.0;
        this.zoomSpeed = 1.2;
    }
    /**
     * @protected
     * @return {?}
     */
    setUpControls() {
        this.controls = new OrbitControls(this.childCameras.first.camera, this.listeningControlElement && this.listeningControlElement.nativeElement);
        this.controls.rotateSpeed = this.rotateSpeed;
        this.controls.zoomSpeed = this.zoomSpeed;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this.controls) {
            return;
        }
        super.ngOnChanges(changes);
        if (changes['rotateSpeed']) {
            this.controls.rotateSpeed = this.rotateSpeed;
        }
        if (changes['zoomSpeed']) {
            this.controls.zoomSpeed = this.zoomSpeed;
        }
    }
}
OrbitControlsComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-orbit-controls',
                template: `
      <ng-content></ng-content>`,
                styles: [":host{display:flex;flex:1}"]
            }] }
];
/** @nocollapse */
OrbitControlsComponent.ctorParameters = () => [
    { type: RendererService },
    { type: RaycasterService }
];
OrbitControlsComponent.propDecorators = {
    rotateSpeed: [{ type: Input }],
    zoomSpeed: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    OrbitControlsComponent.prototype.rotateSpeed;
    /** @type {?} */
    OrbitControlsComponent.prototype.zoomSpeed;
    /**
     * @type {?}
     * @protected
     */
    OrbitControlsComponent.prototype.rendererService;
    /**
     * @type {?}
     * @protected
     */
    OrbitControlsComponent.prototype.raycasterService;
}
