// atft/esm2015/lib/renderer/renderer-canvas.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostListener, ViewChild } from '@angular/core';
import { RendererService } from './renderer.service';
export class RendererCanvasComponent {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        this.rendererService = rendererService;
        // console.log('RendererComponent.constructor');
        this.onResize = this.onResize.bind(this);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // console.log('RendererComponent.ngAfterViewInit');
        this.rendererService.initialize(this.canvas);
    }
    /**
     * The render pane on which the scene is rendered.
     * Currently, only the WebGL renderer with a canvas is used in this
     * implementation, so this property will always be an ElementRef to the
     * underlying <canvas> element.
     *
     * \@example This property can be used to restrict the orbit control (i.e. the
     * area which is listened for mouse move and zoom events) to the rendering pane
     * @return {?}
     */
    get renderPane() {
        return this.canvasRef;
    }
    /**
     * @private
     * @return {?}
     */
    get canvas() {
        return this.canvasRef.nativeElement;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onResize(event) {
        // strange, but single 100% resizing has unexpected behaviour with flex CSS
        // as workaround - resettling to 100 pixels, then to 100%
        this.rendererService.resize(this.canvas, '100px');
        this.rendererService.resize(this.canvas, '100%');
    }
}
RendererCanvasComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-renderer-canvas',
                template: "<canvas #canvas>\r\n  <ng-content></ng-content>\r\n</canvas>\r\n",
                styles: [":host{display:flex;flex:1}canvas{flex:1}"]
            }] }
];
/** @nocollapse */
RendererCanvasComponent.ctorParameters = () => [
    { type: RendererService }
];
RendererCanvasComponent.propDecorators = {
    canvasRef: [{ type: ViewChild, args: ['canvas', { static: true },] }],
    onResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    RendererCanvasComponent.prototype.canvasRef;
    /**
     * @type {?}
     * @private
     */
    RendererCanvasComponent.prototype.rendererService;
}
