// atft/esm5/lib/renderer/renderer-canvas.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostListener, ViewChild } from '@angular/core';
import { RendererService } from './renderer.service';
var RendererCanvasComponent = /** @class */ (function () {
    function RendererCanvasComponent(rendererService) {
        this.rendererService = rendererService;
        // console.log('RendererComponent.constructor');
        this.onResize = this.onResize.bind(this);
    }
    /**
     * @return {?}
     */
    RendererCanvasComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // console.log('RendererComponent.ngAfterViewInit');
        this.rendererService.initialize(this.canvas);
    };
    Object.defineProperty(RendererCanvasComponent.prototype, "renderPane", {
        /**
         * The render pane on which the scene is rendered.
         * Currently, only the WebGL renderer with a canvas is used in this
         * implementation, so this property will always be an ElementRef to the
         * underlying <canvas> element.
         *
         * @example This property can be used to restrict the orbit control (i.e. the
         * area which is listened for mouse move and zoom events) to the rendering pane
         */
        get: /**
         * The render pane on which the scene is rendered.
         * Currently, only the WebGL renderer with a canvas is used in this
         * implementation, so this property will always be an ElementRef to the
         * underlying <canvas> element.
         *
         * \@example This property can be used to restrict the orbit control (i.e. the
         * area which is listened for mouse move and zoom events) to the rendering pane
         * @return {?}
         */
        function () {
            return this.canvasRef;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RendererCanvasComponent.prototype, "canvas", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return this.canvasRef.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    RendererCanvasComponent.prototype.onResize = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // strange, but single 100% resizing has unexpected behaviour with flex CSS
        // as workaround - resettling to 100 pixels, then to 100%
        this.rendererService.resize(this.canvas, '100px');
        this.rendererService.resize(this.canvas, '100%');
    };
    RendererCanvasComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-renderer-canvas',
                    template: "<canvas #canvas>\r\n  <ng-content></ng-content>\r\n</canvas>\r\n",
                    styles: [":host{display:flex;flex:1}canvas{flex:1}"]
                }] }
    ];
    /** @nocollapse */
    RendererCanvasComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    RendererCanvasComponent.propDecorators = {
        canvasRef: [{ type: ViewChild, args: ['canvas', { static: true },] }],
        onResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
    };
    return RendererCanvasComponent;
}());
export { RendererCanvasComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RendererCanvasComponent.prototype.canvasRef;
    /**
     * @type {?}
     * @private
     */
    RendererCanvasComponent.prototype.rendererService;
}
