// atft/esm2015/lib/object/helper/axes-helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
export class AxesHelperComponent extends AbstractObject3D {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        // console.log('AxesHelperComponent.newObject3DInstance');
        return new THREE.AxesHelper(this.size);
    }
}
AxesHelperComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-axes-helper',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => AxesHelperComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
AxesHelperComponent.ctorParameters = () => [
    { type: RendererService }
];
AxesHelperComponent.propDecorators = {
    size: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AxesHelperComponent.prototype.size;
    /**
     * @type {?}
     * @protected
     */
    AxesHelperComponent.prototype.rendererService;
}
