// atft/esm2015/lib/object/helper/grid-helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
export class GridHelperComponent extends AbstractObject3D {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        // console.log('GridHelperComponent.newObject3DInstance');
        return new THREE.GridHelper(this.size, this.divisions);
    }
}
GridHelperComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-grid-helper',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => GridHelperComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
GridHelperComponent.ctorParameters = () => [
    { type: RendererService }
];
GridHelperComponent.propDecorators = {
    size: [{ type: Input }],
    divisions: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    GridHelperComponent.prototype.size;
    /** @type {?} */
    GridHelperComponent.prototype.divisions;
    /**
     * @type {?}
     * @protected
     */
    GridHelperComponent.prototype.rendererService;
}
