// atft/esm2015/lib/object/helper/stats.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * @record
 */
export function StatsResults() { }
if (false) {
    /** @type {?} */
    StatsResults.prototype.ms;
    /** @type {?} */
    StatsResults.prototype.count;
    /** @type {?} */
    StatsResults.prototype.callsPerSecond;
}
export class StatsService {
    constructor() {
        this.ms = 0;
        this.count = 0;
    }
    /**
     * @return {?}
     */
    start() {
        this.lastTime = Date.now();
    }
    /**
     * @return {?}
     */
    end() {
        /** @type {?} */
        const deltaMs = Date.now() - this.lastTime;
        this.ms += deltaMs;
        this.count++;
    }
    /**
     * @return {?}
     */
    getStats() {
        return {
            ms: this.ms,
            count: this.count,
            callsPerSecond: Math.round(this.count / (this.ms / 1000))
        };
    }
    /**
     * @return {?}
     */
    reset() {
        this.ms = 0;
        this.count = 0;
    }
}
StatsService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.lastTime;
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.ms;
    /**
     * @type {?}
     * @private
     */
    StatsService.prototype.count;
}
