// atft/esm5/lib/object/connector/abstract-connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
/**
 * @abstract
 * @template T
 */
var AbstractConnector = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractConnector, _super);
    function AbstractConnector() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractConnector.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var mesh = this.createConnectorObject();
        this.watchObjects();
        return mesh;
    };
    /**
     * @private
     * @return {?}
     */
    AbstractConnector.prototype.watchObjects = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.source.changed.subscribe((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.updateLineGeometry();
        }));
        this.target.changed.subscribe((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.updateLineGeometry();
        }));
    };
    /**
     * @protected
     * @return {?}
     */
    AbstractConnector.prototype.getLineGeometry = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var geo = new THREE.Geometry();
        if (!this.source || !this.target) {
            throw new Error('AbstractConnector: source or target inputs are missing!');
        }
        geo.vertices.push(this.source.getObject().position);
        geo.vertices.push(this.target.getObject().position);
        return geo;
    };
    AbstractConnector.propDecorators = {
        source: [{ type: Input }],
        target: [{ type: Input }]
    };
    return AbstractConnector;
}(AbstractObject3D));
export { AbstractConnector };
if (false) {
    /** @type {?} */
    AbstractConnector.prototype.source;
    /** @type {?} */
    AbstractConnector.prototype.target;
    /**
     * Create line mesh
     * @abstract
     * @return {?}
     */
    AbstractConnector.prototype.createConnectorObject = function () { };
    /**
     * If at least one line end (source or target object)  changed, then line geoetry should be updated as well
     * // TODO: Calculate only when source/target positions were changed
     * @abstract
     * @return {?}
     */
    AbstractConnector.prototype.updateLineGeometry = function () { };
}
