// atft/esm5/lib/object/connector/line-connector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractConnector } from './abstract-connector';
import { appliedColor } from '../../util/applied-color';
import { RendererService } from '../../renderer/renderer.service';
var LineConnectorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LineConnectorComponent, _super);
    function LineConnectorComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.materialColor = 0xffff00;
        return _this;
    }
    /**
     * @return {?}
     */
    LineConnectorComponent.prototype.createConnectorObject = /**
     * @return {?}
     */
    function () {
        this.geometry = this.getLineGeometry();
        /** @type {?} */
        var material = new THREE.LineBasicMaterial({
            color: appliedColor(this.materialColor),
            /**
             * NOTE: linewidth=1: Due to limitations of the OpenGL Core Profile
             * with the WebGL renderer on most platforms linewidth will always be 1
             * regardless of the set value.
             */
            linewidth: 1
        });
        /** @type {?} */
        var line = new THREE.Line(this.geometry, material);
        return line;
    };
    /**
     * @return {?}
     */
    LineConnectorComponent.prototype.updateLineGeometry = /**
     * @return {?}
     */
    function () {
        this.geometry.verticesNeedUpdate = true;
        this.rendererService.render();
    };
    LineConnectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-line-connector',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return LineConnectorComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    LineConnectorComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    LineConnectorComponent.propDecorators = {
        materialColor: [{ type: Input }]
    };
    return LineConnectorComponent;
}(AbstractConnector));
export { LineConnectorComponent };
if (false) {
    /** @type {?} */
    LineConnectorComponent.prototype.materialColor;
    /**
     * @type {?}
     * @private
     */
    LineConnectorComponent.prototype.geometry;
    /**
     * @type {?}
     * @protected
     */
    LineConnectorComponent.prototype.rendererService;
}
