// atft/esm5/lib/object/loader/obj-loader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { AbstractModelLoader } from './abstract-model-loader';
import { OBJLoader } from 'three/examples/jsm/loaders/OBJLoader';
import { MTLLoader } from 'three/examples/jsm/loaders/MTLLoader';
import { RendererService } from '../../renderer/renderer.service';
/**
 * Component for employing THREE.OBJLoader to load [Wavefront *.obj files][1].
 *
 * [1]: https://en.wikipedia.org/wiki/Wavefront_.obj_file
 */
var ObjLoaderComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ObjLoaderComponent, _super);
    function ObjLoaderComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.loader = new OBJLoader();
        _this.mtlLoader = new MTLLoader();
        return _this;
    }
    Object.defineProperty(ObjLoaderComponent.prototype, "texturePath", {
        /**
         * Path relative to which resources and textures within the loaded obj file are loaded.
         * @deprecated Rather use resourcePath
         */
        set: /**
         * Path relative to which resources and textures within the loaded obj file are loaded.
         * @deprecated Rather use resourcePath
         * @param {?} newTexturePath
         * @return {?}
         */
        function (newTexturePath) {
            this.resourcePath = newTexturePath;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @protected
     * @return {?}
     */
    ObjLoaderComponent.prototype.loadLazyObject = /**
     * @protected
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var preloadingStep;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                // Preloading step for the material
                preloadingStep = new Promise((/**
                 * @param {?} resolve
                 * @param {?} reject
                 * @return {?}
                 */
                function (resolve, reject) {
                    if (_this.material === undefined) {
                        // No preloading necessary
                        resolve();
                    }
                    else {
                        // TODO: If typings of mtlLoader are included in the Three.js NPM
                        // package, remove this 'any' cast.
                        ((/** @type {?} */ (_this.mtlLoader))).setResourcePath(_this.resourcePath);
                        _this.mtlLoader.load(_this.material, (/**
                         * @param {?} materialCreator
                         * @return {?}
                         */
                        function (materialCreator) {
                            materialCreator.preload();
                            _this.loader.setMaterials(materialCreator);
                            resolve();
                        }), undefined, reject);
                    }
                }));
                // Await preloading and load final model
                return [2 /*return*/, preloadingStep.then((/**
                     * @return {?}
                     */
                    function () {
                        return new Promise((/**
                         * @param {?} resolve
                         * @param {?} reject
                         * @return {?}
                         */
                        function (resolve, reject) {
                            _this.loader.load(_this.model, (/**
                             * @param {?} model
                             * @return {?}
                             */
                            function (model) {
                                resolve(model);
                            }), undefined, reject);
                        }));
                    }))];
            });
        });
    };
    ObjLoaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-obj-loader',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ObjLoaderComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    ObjLoaderComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    ObjLoaderComponent.propDecorators = {
        material: [{ type: Input }],
        texturePath: [{ type: Input }],
        resourcePath: [{ type: Input }]
    };
    return ObjLoaderComponent;
}(AbstractModelLoader));
export { ObjLoaderComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ObjLoaderComponent.prototype.loader;
    /**
     * @type {?}
     * @private
     */
    ObjLoaderComponent.prototype.mtlLoader;
    /** @type {?} */
    ObjLoaderComponent.prototype.material;
    /**
     * Path relative to which resources and textures within the loaded obj file are loaded.
     * @type {?}
     */
    ObjLoaderComponent.prototype.resourcePath;
    /**
     * @type {?}
     * @protected
     */
    ObjLoaderComponent.prototype.rendererService;
}
