// atft/esm5/lib/object/abstract-lazy-object-3d.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AbstractObject3D } from './abstract-object-3d';
import * as THREE from 'three';
/**
 * @abstract
 */
var /**
 * @abstract
 */
AbstractLazyObject3D = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractLazyObject3D, _super);
    function AbstractLazyObject3D() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Flag to signal whether the parent class instance AbstractObject3D called the
         * overwritten method {\@link AbstractModelLoader#afterInit} yet.
         *
         * Unless that method was called, no methods and properties of {\@link AbstractObject3D}
         * may be safely accessed, especially {\@link AbstractObject3D#addChild} and
         * {\@link AbstractObject3D.renderer}.
         */
        _this.parentInitialized = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractLazyObject3D.prototype.afterInit = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.afterInit.call(this);
        this.parentInitialized = true;
        this.startLoading();
    };
    /**
     * @protected
     * @return {?}
     */
    AbstractLazyObject3D.prototype.startLoading = /**
     * @protected
     * @return {?}
     */
    function () {
        var _this = this;
        // console.log('AbstractLazyObject3D.startLoading');
        // Trigger model acquisition now that the parent has been initialized.
        this.loadLazyObject().then((/**
         * @param {?} obj
         * @return {?}
         */
        function (obj) {
            // console.log('AbstractLazyObject3D loaded');
            // remove old if exists
            if (_this.lazyObject) {
                _super.prototype.removeChild.call(_this, _this.lazyObject);
            }
            // add lazy object to scene
            _this.lazyObject = obj;
            _super.prototype.addChild.call(_this, obj);
            _this.rendererService.render();
        })).catch((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
        }));
    };
    /**
     * @return {?}
     */
    AbstractLazyObject3D.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.lazyObject) {
            _super.prototype.removeChild.call(this, this.lazyObject);
        }
    };
    /**
     * @protected
     * @return {?}
     */
    AbstractLazyObject3D.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        // Just empty object (holder of lazy object)
        return new THREE.Object3D();
    };
    return AbstractLazyObject3D;
}(AbstractObject3D));
/**
 * @abstract
 */
export { AbstractLazyObject3D };
if (false) {
    /**
     * Flag to signal whether the parent class instance AbstractObject3D called the
     * overwritten method {\@link AbstractModelLoader#afterInit} yet.
     *
     * Unless that method was called, no methods and properties of {\@link AbstractObject3D}
     * may be safely accessed, especially {\@link AbstractObject3D#addChild} and
     * {\@link AbstractObject3D.renderer}.
     * @type {?}
     * @private
     */
    AbstractLazyObject3D.prototype.parentInitialized;
    /**
     * This is reference to lazy loaded Object3D (async after init)
     * @type {?}
     * @protected
     */
    AbstractLazyObject3D.prototype.lazyObject;
    /**
     * Abstract method for lazy loading
     *
     * @abstract
     * @protected
     * @return {?}
     */
    AbstractLazyObject3D.prototype.loadLazyObject = function () { };
}
