// atft/esm5/lib/object/scene.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from './abstract-object-3d';
import { appliedColor } from '../util/applied-color';
import { RendererService } from '../renderer/renderer.service';
var SceneComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SceneComponent, _super);
    function SceneComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.background = 0xffffff;
        _this.fog = false;
        _this.fogColor = 0xa0a0a0;
        _this.fogNear = 10;
        _this.fogFar = 500;
        // TODO: directive?
        rendererService.setScene(_this);
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SceneComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var scene = new THREE.Scene();
        scene.background = new THREE.Color(appliedColor(this.background));
        if (this.fog === true) {
            scene.fog = new THREE.Fog(appliedColor(this.fogColor), this.fogNear, this.fogFar);
        }
        return scene;
    };
    SceneComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-scene',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return SceneComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    SceneComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    SceneComponent.propDecorators = {
        background: [{ type: Input }],
        fog: [{ type: Input }],
        fogColor: [{ type: Input }],
        fogNear: [{ type: Input }],
        fogFar: [{ type: Input }]
    };
    return SceneComponent;
}(AbstractObject3D));
export { SceneComponent };
if (false) {
    /** @type {?} */
    SceneComponent.prototype.background;
    /** @type {?} */
    SceneComponent.prototype.fog;
    /** @type {?} */
    SceneComponent.prototype.fogColor;
    /** @type {?} */
    SceneComponent.prototype.fogNear;
    /** @type {?} */
    SceneComponent.prototype.fogFar;
    /**
     * @type {?}
     * @protected
     */
    SceneComponent.prototype.rendererService;
}
