// atft/esm2015/lib/object/light/hemisphere-light.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from '../abstract-object-3d';
import { appliedColor } from '../../util/applied-color';
import { RendererService } from '../../renderer/renderer.service';
export class HemisphereLightComponent extends AbstractObject3D {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.skyColor = 0xffffff;
        this.groundColor = 0x444444;
        this.intensity = 1;
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        /** @type {?} */
        const light = new THREE.HemisphereLight(appliedColor(this.skyColor), appliedColor(this.groundColor), appliedColor(this.intensity));
        return light;
    }
}
HemisphereLightComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-hemisphere-light',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => HemisphereLightComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
HemisphereLightComponent.ctorParameters = () => [
    { type: RendererService }
];
HemisphereLightComponent.propDecorators = {
    skyColor: [{ type: Input }],
    groundColor: [{ type: Input }],
    intensity: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HemisphereLightComponent.prototype.skyColor;
    /** @type {?} */
    HemisphereLightComponent.prototype.groundColor;
    /** @type {?} */
    HemisphereLightComponent.prototype.intensity;
    /**
     * @type {?}
     * @protected
     */
    HemisphereLightComponent.prototype.rendererService;
}
