// atft/esm5/lib/object/mesh/abstract-css3d-mesh.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { calculateSize } from '../../util';
/**
 * @abstract
 */
var AbstractCss3dMesh = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractCss3dMesh, _super);
    function AbstractCss3dMesh() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.width = 1.0;
        _this.height = 1.0;
        // 1280 / 720 = 1.777777778
        _this.resolutionX = 1280;
        _this.resolutionY = 720;
        _this.widthSegments = 1;
        _this.heightSegments = 1;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractCss3dMesh.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var geometry = new THREE.PlaneBufferGeometry(this.width, this.height, this.widthSegments, this.heightSegments);
        /** @type {?} */
        var material = this.getMaterial();
        /** @type {?} */
        var mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        // Create and scale css3d object:
        /** @type {?} */
        var meshSize = calculateSize(mesh);
        /** @type {?} */
        var css3dObject = this.createCss3dObject();
        this.scaleCss3d(css3dObject, meshSize);
        // append to mesh:
        mesh.add(css3dObject);
        return mesh;
    };
    /**
     * @protected
     * @param {?} group
     * @param {?} max
     * @return {?}
     */
    AbstractCss3dMesh.prototype.scaleCss3d = /**
     * @protected
     * @param {?} group
     * @param {?} max
     * @return {?}
     */
    function (group, max) {
        /** @type {?} */
        var box = new THREE.Vector3(this.resolutionX, this.resolutionY, 0);
        /** @type {?} */
        var scaleX = max.x / box.x;
        /** @type {?} */
        var scaleY = max.y / box.y;
        /** @type {?} */
        var scaleZ = max.z / box.z;
        group.scale.set((scaleX < 1 ? scaleX : 1), (scaleY < 1 ? scaleY : 1), (scaleZ < 1 ? scaleZ : 1));
    };
    /**
     * @protected
     * @return {?}
     */
    AbstractCss3dMesh.prototype.getMaterial = /**
     * @protected
     * @return {?}
     */
    function () {
        // TODO: Renderer clear color
        /** @type {?} */
        var transparentMaterial = new THREE.MeshBasicMaterial({
            color: 0x000000,
            opacity: 0,
            side: THREE.DoubleSide
        });
        return transparentMaterial;
    };
    AbstractCss3dMesh.propDecorators = {
        width: [{ type: Input }],
        height: [{ type: Input }],
        resolutionX: [{ type: Input }],
        resolutionY: [{ type: Input }]
    };
    return AbstractCss3dMesh;
}(AbstractMesh));
export { AbstractCss3dMesh };
if (false) {
    /** @type {?} */
    AbstractCss3dMesh.prototype.width;
    /** @type {?} */
    AbstractCss3dMesh.prototype.height;
    /** @type {?} */
    AbstractCss3dMesh.prototype.resolutionX;
    /** @type {?} */
    AbstractCss3dMesh.prototype.resolutionY;
    /**
     * @type {?}
     * @private
     */
    AbstractCss3dMesh.prototype.widthSegments;
    /**
     * @type {?}
     * @private
     */
    AbstractCss3dMesh.prototype.heightSegments;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    AbstractCss3dMesh.prototype.createCss3dObject = function () { };
}
