// atft/esm5/lib/object/mesh/abstract-mesh-3d.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { appliedMaterial } from '../../util';
/**
 * @abstract
 */
var AbstractMesh = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractMesh, _super);
    function AbstractMesh() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.materialColor = 0x5DADE2;
        _this.castShadow = true;
        _this.receiveShadow = true;
        _this.depthWrite = true;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractMesh.prototype.getMaterial = /**
     * @protected
     * @return {?}
     */
    function () {
        return appliedMaterial(this.materialColor, this.material, this.depthWrite);
    };
    /**
     * @protected
     * @param {?} mesh
     * @return {?}
     */
    AbstractMesh.prototype.applyShadowProps = /**
     * @protected
     * @param {?} mesh
     * @return {?}
     */
    function (mesh) {
        mesh.castShadow = this.castShadow;
        mesh.receiveShadow = this.receiveShadow;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AbstractMesh.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        _super.prototype.ngOnChanges.call(this, changes);
        if (!this.getObject()) {
            return;
        }
        /** @type {?} */
        var mustRerender = false;
        if (['material', 'materialColor', 'depthWrite'].some((/**
         * @param {?} propName
         * @return {?}
         */
        function (propName) { return propName in changes; }))) {
            this.applyMaterial();
            mustRerender = true;
        }
        if (mustRerender) {
            this.rendererService.render();
        }
    };
    /**
     * @return {?}
     */
    AbstractMesh.prototype.applyMaterial = /**
     * @return {?}
     */
    function () {
        this.getObject().material = this.getMaterial();
    };
    AbstractMesh.propDecorators = {
        material: [{ type: Input }],
        materialColor: [{ type: Input }],
        castShadow: [{ type: Input }],
        receiveShadow: [{ type: Input }],
        depthWrite: [{ type: Input }]
    };
    return AbstractMesh;
}(AbstractObject3D));
export { AbstractMesh };
if (false) {
    /** @type {?} */
    AbstractMesh.prototype.material;
    /** @type {?} */
    AbstractMesh.prototype.materialColor;
    /** @type {?} */
    AbstractMesh.prototype.castShadow;
    /** @type {?} */
    AbstractMesh.prototype.receiveShadow;
    /** @type {?} */
    AbstractMesh.prototype.depthWrite;
}
