// atft/esm5/lib/object/mesh/css3d-video-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractObject3D } from '../abstract-object-3d';
import { CSS3DObject } from 'three/examples/jsm/renderers/CSS3DRenderer';
import { AbstractCss3dMesh } from './abstract-css3d-mesh';
import { RendererService } from '../../renderer/renderer.service';
var Css3dVideoMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(Css3dVideoMeshComponent, _super);
    function Css3dVideoMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.video = document.createElement('video');
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    Css3dVideoMeshComponent.prototype.createCss3dObject = /**
     * @protected
     * @return {?}
     */
    function () {
        // 1. Create DIV
        /** @type {?} */
        var div = document.createElement('div');
        div.style.width = this.resolutionX + 'px';
        div.style.height = this.resolutionY + 'px';
        div.style.backgroundColor = '#000000';
        div.id = 'myWrapper';
        // 2. Create Video Player with source
        this.video.autoplay = false;
        this.video.loop = true;
        this.video.muted = true;
        /** @type {?} */
        var source = document.createElement('source');
        source.src = this.videoSrc;
        source.type = 'video/mp4';
        this.video.appendChild(source);
        div.appendChild(this.video);
        // 3. Create CSS3DObject
        /** @type {?} */
        var object = new CSS3DObject(div);
        return object;
    };
    /**
     * @return {?}
     */
    Css3dVideoMeshComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterViewInit.call(this);
        this.video.load();
        this.video.addEventListener('canplay', (/**
         * @return {?}
         */
        function () {
            console.log('EVENT');
            _this.video.play();
        }));
    };
    Css3dVideoMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-css3d-video-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return Css3dVideoMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    Css3dVideoMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    Css3dVideoMeshComponent.propDecorators = {
        videoSrc: [{ type: Input }]
    };
    return Css3dVideoMeshComponent;
}(AbstractCss3dMesh));
export { Css3dVideoMeshComponent };
if (false) {
    /** @type {?} */
    Css3dVideoMeshComponent.prototype.videoSrc;
    /** @type {?} */
    Css3dVideoMeshComponent.prototype.video;
    /**
     * @type {?}
     * @protected
     */
    Css3dVideoMeshComponent.prototype.rendererService;
}
