// atft/esm5/lib/object/mesh/cylinder-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var CylinderMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CylinderMeshComponent, _super);
    function CylinderMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.radiusTop = 1.0;
        _this.radiusBottom = 1.0;
        _this.height = 1.0;
        _this.radialSegments = 8;
        _this.heightSegments = 1;
        _this.openEnded = false;
        _this.thetaStart = 0.0;
        _this.thetaLength = 2 * Math.PI;
        return _this;
        // console.log('CylinderMeshComponent.constructor');
    }
    /**
     * @protected
     * @return {?}
     */
    CylinderMeshComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        // console.log('CylinderMeshComponent.newObject3DInstance');
        /** @type {?} */
        var geometry = new THREE.CylinderGeometry(this.radiusTop, this.radiusBottom, this.height, this.radialSegments, this.heightSegments, this.openEnded, this.thetaStart, this.thetaLength);
        /** @type {?} */
        var material = this.getMaterial();
        /** @type {?} */
        var mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        return mesh;
    };
    CylinderMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-cylinder-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return CylinderMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    CylinderMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    CylinderMeshComponent.propDecorators = {
        radiusTop: [{ type: Input }],
        radiusBottom: [{ type: Input }],
        height: [{ type: Input }],
        radialSegments: [{ type: Input }],
        heightSegments: [{ type: Input }],
        openEnded: [{ type: Input }],
        thetaStart: [{ type: Input }],
        thetaLength: [{ type: Input }]
    };
    return CylinderMeshComponent;
}(AbstractMesh));
export { CylinderMeshComponent };
if (false) {
    /** @type {?} */
    CylinderMeshComponent.prototype.radiusTop;
    /** @type {?} */
    CylinderMeshComponent.prototype.radiusBottom;
    /** @type {?} */
    CylinderMeshComponent.prototype.height;
    /** @type {?} */
    CylinderMeshComponent.prototype.radialSegments;
    /** @type {?} */
    CylinderMeshComponent.prototype.heightSegments;
    /** @type {?} */
    CylinderMeshComponent.prototype.openEnded;
    /** @type {?} */
    CylinderMeshComponent.prototype.thetaStart;
    /** @type {?} */
    CylinderMeshComponent.prototype.thetaLength;
    /**
     * @type {?}
     * @protected
     */
    CylinderMeshComponent.prototype.rendererService;
}
