// atft/esm5/lib/object/mesh/frame-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var FrameMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FrameMeshComponent, _super);
    function FrameMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        _this.thickness = 2;
        _this.sizeX = 20;
        _this.sizeY = 20;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FrameMeshComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var halfX = this.sizeX / 2.0;
        /** @type {?} */
        var halfY = this.sizeY / 2.0;
        /** @type {?} */
        var t = this.thickness;
        /** @type {?} */
        var shape = new THREE.Shape();
        shape.moveTo(-halfX, halfY);
        shape.lineTo(-halfX - t, halfY + t);
        shape.lineTo(halfX + t, halfY + t);
        shape.lineTo(halfX + t, -halfY - t);
        shape.lineTo(-halfX - t, -halfY - t);
        shape.lineTo(-halfX - t, halfY + t);
        shape.lineTo(-halfX, halfY);
        shape.lineTo(-halfX, -halfY);
        shape.lineTo(halfX, -halfY);
        shape.lineTo(halfX, halfY);
        shape.lineTo(-halfX, halfY);
        /** @type {?} */
        var geometry = new THREE.ShapeBufferGeometry(shape);
        /** @type {?} */
        var material = this.getMaterial();
        /** @type {?} */
        var mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        return mesh;
    };
    FrameMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-frame-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return FrameMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    FrameMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    FrameMeshComponent.propDecorators = {
        thickness: [{ type: Input }],
        sizeX: [{ type: Input }],
        sizeY: [{ type: Input }]
    };
    return FrameMeshComponent;
}(AbstractMesh));
export { FrameMeshComponent };
if (false) {
    /** @type {?} */
    FrameMeshComponent.prototype.thickness;
    /** @type {?} */
    FrameMeshComponent.prototype.sizeX;
    /** @type {?} */
    FrameMeshComponent.prototype.sizeY;
    /**
     * @type {?}
     * @protected
     */
    FrameMeshComponent.prototype.rendererService;
}
