// atft/esm5/lib/object/mesh/sphere-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var SphereMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SphereMeshComponent, _super);
    function SphereMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SphereMeshComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        // console.log('SphereMeshComponent.newObject3DInstance');
        /** @type {?} */
        var geometry = new THREE.SphereGeometry(this.radius, this.widthSegments, this.hightSegments);
        /** @type {?} */
        var material = this.getMaterial();
        /** @type {?} */
        var mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        return mesh;
    };
    SphereMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-sphere-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return SphereMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    SphereMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    SphereMeshComponent.propDecorators = {
        radius: [{ type: Input }],
        widthSegments: [{ type: Input }],
        hightSegments: [{ type: Input }]
    };
    return SphereMeshComponent;
}(AbstractMesh));
export { SphereMeshComponent };
if (false) {
    /** @type {?} */
    SphereMeshComponent.prototype.radius;
    /** @type {?} */
    SphereMeshComponent.prototype.widthSegments;
    /** @type {?} */
    SphereMeshComponent.prototype.hightSegments;
    /**
     * @type {?}
     * @protected
     */
    SphereMeshComponent.prototype.rendererService;
}
