// atft/esm5/lib/object/mesh/torus-mesh.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractMesh } from './abstract-mesh-3d';
import { AbstractObject3D } from '../abstract-object-3d';
import { RendererService } from '../../renderer/renderer.service';
var TorusMeshComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TorusMeshComponent, _super);
    function TorusMeshComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        /**
         * Radius of the torus, from the center of the torus to the center of the tube
         */
        _this.radius = 0.4;
        _this.radialSegments = 8;
        _this.tubularSegments = 6;
        _this.arc = Math.PI * 2;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TorusMeshComponent.prototype.newObject3DInstance = /**
     * @protected
     * @return {?}
     */
    function () {
        this.radius *= 1;
        this.tube *= 1;
        this.radialSegments *= 1;
        this.tubularSegments *= 1;
        /** @type {?} */
        var geometry = new THREE.TorusGeometry(this.radius, this.tube, this.radialSegments, this.tubularSegments);
        /** @type {?} */
        var material = this.getMaterial();
        /** @type {?} */
        var mesh = new THREE.Mesh(geometry, material);
        this.applyShadowProps(mesh);
        return mesh;
    };
    TorusMeshComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-torus-mesh',
                    providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TorusMeshComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    TorusMeshComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    TorusMeshComponent.propDecorators = {
        radius: [{ type: Input }],
        tube: [{ type: Input }],
        radialSegments: [{ type: Input }],
        tubularSegments: [{ type: Input }],
        arc: [{ type: Input }]
    };
    return TorusMeshComponent;
}(AbstractMesh));
export { TorusMeshComponent };
if (false) {
    /**
     * Radius of the torus, from the center of the torus to the center of the tube
     * @type {?}
     */
    TorusMeshComponent.prototype.radius;
    /**
     * Radius of the tube.
     * @type {?}
     */
    TorusMeshComponent.prototype.tube;
    /** @type {?} */
    TorusMeshComponent.prototype.radialSegments;
    /** @type {?} */
    TorusMeshComponent.prototype.tubularSegments;
    /** @type {?} */
    TorusMeshComponent.prototype.arc;
    /**
     * @type {?}
     * @protected
     */
    TorusMeshComponent.prototype.rendererService;
}
