// atft/esm5/lib/camera/perspective-camera.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { AbstractCamera } from './abstract-camera';
import * as THREE from 'three';
import { RendererService } from '../renderer/renderer.service';
var PerspectiveCameraComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PerspectiveCameraComponent, _super);
    function PerspectiveCameraComponent(rendererService) {
        var _this = _super.call(this, rendererService) || this;
        _this.rendererService = rendererService;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    PerspectiveCameraComponent.prototype.createCamera = /**
     * @protected
     * @return {?}
     */
    function () {
        // console.log('PerspectiveCameraComponent.createCamera');
        // let aspectRatio = undefined; // Updated later
        this.camera = new THREE.PerspectiveCamera(this.fov, undefined, this.near, this.far);
    };
    /**
     * @param {?} aspect
     * @return {?}
     */
    PerspectiveCameraComponent.prototype.updateAspectRatio = /**
     * @param {?} aspect
     * @return {?}
     */
    function (aspect) {
        // console.log('PerspectiveCameraComponent.updateAspectRatio: ' + aspect);
        this.camera.aspect = aspect;
        this.camera.updateProjectionMatrix();
    };
    PerspectiveCameraComponent.decorators = [
        { type: Component, args: [{
                    selector: 'atft-perspective-camera',
                    providers: [{ provide: AbstractCamera, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return PerspectiveCameraComponent; })) }],
                    template: '<ng-content></ng-content>'
                }] }
    ];
    /** @nocollapse */
    PerspectiveCameraComponent.ctorParameters = function () { return [
        { type: RendererService }
    ]; };
    PerspectiveCameraComponent.propDecorators = {
        fov: [{ type: Input }],
        near: [{ type: Input }],
        far: [{ type: Input }]
    };
    return PerspectiveCameraComponent;
}(AbstractCamera));
export { PerspectiveCameraComponent };
if (false) {
    /** @type {?} */
    PerspectiveCameraComponent.prototype.fov;
    /** @type {?} */
    PerspectiveCameraComponent.prototype.near;
    /** @type {?} */
    PerspectiveCameraComponent.prototype.far;
    /**
     * @type {?}
     * @protected
     */
    PerspectiveCameraComponent.prototype.rendererService;
}
