// atft/esm2015/lib/object/abstract-lazy-object-3d.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AbstractObject3D } from './abstract-object-3d';
import * as THREE from 'three';
/**
 * @abstract
 */
export class AbstractLazyObject3D extends AbstractObject3D {
    constructor() {
        super(...arguments);
        /**
         * Flag to signal whether the parent class instance AbstractObject3D called the
         * overwritten method {\@link AbstractModelLoader#afterInit} yet.
         *
         * Unless that method was called, no methods and properties of {\@link AbstractObject3D}
         * may be safely accessed, especially {\@link AbstractObject3D#addChild} and
         * {\@link AbstractObject3D.renderer}.
         */
        this.parentInitialized = false;
    }
    /**
     * @protected
     * @return {?}
     */
    afterInit() {
        super.afterInit();
        this.parentInitialized = true;
        this.startLoading();
    }
    /**
     * @protected
     * @return {?}
     */
    startLoading() {
        // console.log('AbstractLazyObject3D.startLoading');
        // Trigger model acquisition now that the parent has been initialized.
        this.loadLazyObject().then((/**
         * @param {?} obj
         * @return {?}
         */
        obj => {
            // console.log('AbstractLazyObject3D loaded');
            // remove old if exists
            if (this.lazyObject) {
                super.removeChild(this.lazyObject);
            }
            // add lazy object to scene
            this.lazyObject = obj;
            super.addChild(obj);
            this.rendererService.render();
        })).catch((/**
         * @param {?} err
         * @return {?}
         */
        err => {
            console.error(err);
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.lazyObject) {
            super.removeChild(this.lazyObject);
        }
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        // Just empty object (holder of lazy object)
        return new THREE.Object3D();
    }
}
if (false) {
    /**
     * Flag to signal whether the parent class instance AbstractObject3D called the
     * overwritten method {\@link AbstractModelLoader#afterInit} yet.
     *
     * Unless that method was called, no methods and properties of {\@link AbstractObject3D}
     * may be safely accessed, especially {\@link AbstractObject3D#addChild} and
     * {\@link AbstractObject3D.renderer}.
     * @type {?}
     * @private
     */
    AbstractLazyObject3D.prototype.parentInitialized;
    /**
     * This is reference to lazy loaded Object3D (async after init)
     * @type {?}
     * @protected
     */
    AbstractLazyObject3D.prototype.lazyObject;
    /**
     * Abstract method for lazy loading
     *
     * @abstract
     * @protected
     * @return {?}
     */
    AbstractLazyObject3D.prototype.loadLazyObject = function () { };
}
