// atft/esm2015/lib/object/scene.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import * as THREE from 'three';
import { AbstractObject3D } from './abstract-object-3d';
import { appliedColor } from '../util/applied-color';
import { RendererService } from '../renderer/renderer.service';
export class SceneComponent extends AbstractObject3D {
    /**
     * @param {?} rendererService
     */
    constructor(rendererService) {
        super(rendererService);
        this.rendererService = rendererService;
        this.background = 0xffffff;
        this.fog = false;
        this.fogColor = 0xa0a0a0;
        this.fogNear = 10;
        this.fogFar = 500;
        // TODO: directive?
        rendererService.setScene(this);
    }
    /**
     * @protected
     * @return {?}
     */
    newObject3DInstance() {
        /** @type {?} */
        const scene = new THREE.Scene();
        scene.background = new THREE.Color(appliedColor(this.background));
        if (this.fog === true) {
            scene.fog = new THREE.Fog(appliedColor(this.fogColor), this.fogNear, this.fogFar);
        }
        return scene;
    }
}
SceneComponent.decorators = [
    { type: Component, args: [{
                selector: 'atft-scene',
                providers: [{ provide: AbstractObject3D, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => SceneComponent)) }],
                template: '<ng-content></ng-content>'
            }] }
];
/** @nocollapse */
SceneComponent.ctorParameters = () => [
    { type: RendererService }
];
SceneComponent.propDecorators = {
    background: [{ type: Input }],
    fog: [{ type: Input }],
    fogColor: [{ type: Input }],
    fogNear: [{ type: Input }],
    fogFar: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SceneComponent.prototype.background;
    /** @type {?} */
    SceneComponent.prototype.fog;
    /** @type {?} */
    SceneComponent.prototype.fogColor;
    /** @type {?} */
    SceneComponent.prototype.fogNear;
    /** @type {?} */
    SceneComponent.prototype.fogFar;
    /**
     * @type {?}
     * @protected
     */
    SceneComponent.prototype.rendererService;
}
